File: src/MixNetFx/FacessoData/DataClasses/ParameterClasses/ProductionPeriods.vb
Imports ActiveDev
Imports System.Collections.ObjectModel
Imports System.Globalization

Public Class ProductionPeriod
    Inherits Collection(Of ProductionPeriodItem)

    Private _StartDate As Date
    Private _EndDate As Date
    Private _ShiftParameters As ShiftParameters

    Public Sub New(ByVal CertainDate As Date, ByVal CertainShift As Byte)
        MyBase.New()
        CertainDate = CertainDate.Date
        _StartDate = CertainDate
        _EndDate = CertainDate
        Dim locProductionPeriodItem As New ProductionPeriodItem(CertainDate, CertainShift)
        Me.Add(locProductionPeriodItem)
    End Sub

    ''' <summary>
    ''' Erstellt einen neuen Produktionsdaten-Auswertungszeitraum über alle Schichten im entsprechenden Zeitraum
    ''' </summary>
    ''' <param name="Startdate">Das Anfangsdatum des Auswertungszeitraums</param>
    ''' <param name="EndDate">Das Enddatum des Auswertungszeitraums</param>
    ''' <remarks></remarks>
    Public Sub New(ByVal Startdate As Date, ByVal EndDate As Date)
        MyBase.new()
        Startdate = Startdate.Date
        EndDate = EndDate.Date
        _StartDate = Startdate
        _EndDate = EndDate
        For locDouble As Double = Startdate.ToOADate To EndDate.ToOADate
            For locShift As Byte = 1 To 4
                Dim locProductionPeriodItem As New ProductionPeriodItem(Date.FromOADate(locDouble), locShift)
                Me.Add(locProductionPeriodItem)
            Next
        Next
    End Sub

    ''' <summary>
    ''' Erstellt einen neuen Produktionsdaten-Auswertungszeitraum über eine bestimmte Schicht im entsprechenden Zeitraum
    ''' </summary>
    ''' <param name="Startdate">Das Anfangsdatum des Auswertungszeitraums</param>
    ''' <param name="EndDate">Das Enddatum des Auswertungszeitraums</param>
    ''' <param name="Shift">Die zu berücksichtigende Schicht</param>
    ''' <remarks></remarks>
    Public Sub New(ByVal Startdate As Date, ByVal EndDate As Date, ByVal Shift As Byte)
        Me.New(Startdate, EndDate, Shift, Shift, 1)
    End Sub

    ''' <summary>
    ''' Erstellt einen neuen Produktionsdaten-Auswertungszeitraum über eine oder mehrere, im Bedarfsfall alle X Tage wechselnde Schicht im entsprechenden Zeitraum
    ''' </summary>
    ''' <param name="DateRange">Bestimmt den Zeitraum.</param>
    ''' <param name="Shifts">Bestimmt die Schichten und, falls gewünscht, nach wievielen Tagen diese wechseln sollen.</param>
    ''' <remarks></remarks>
    Public Sub New(ByVal DateRange As DateRangeParameter, ByVal Shifts As ShiftParameters)
        MyBase.New()
        _StartDate = DateRange.StartDate.Date
        _EndDate = DateRange.EndDate.Date
        _ShiftParameters = Shifts

        Dim locShiftDaysChangeCount As Integer = 0
        Dim locShiftChangeFlag As Boolean = False
        For locDouble As Double = _StartDate.ToOADate To _EndDate.ToOADate
            If Not Shifts.AlternateShifts Then
                If Shifts.ConsiderShift1 Then Me.Add(New ProductionPeriodItem(Date.FromOADate(locDouble), 1))
                If Shifts.ConsiderShift2 Then Me.Add(New ProductionPeriodItem(Date.FromOADate(locDouble), 2))
                If Shifts.ConsiderShift3 Then Me.Add(New ProductionPeriodItem(Date.FromOADate(locDouble), 3))
                If Shifts.ConsiderShift4 Then Me.Add(New ProductionPeriodItem(Date.FromOADate(locDouble), 4))
            Else
                Dim locProductionPeriodItem As New ProductionPeriodItem(Date.FromOADate(locDouble), CByte(IIf(locShiftChangeFlag, Shifts.AlternatingFirstShift, Shifts.AlternatingSecondShift)))
                locShiftDaysChangeCount += 1
                If locShiftDaysChangeCount > 0 Then
                    If locShiftDaysChangeCount = Shifts.DaysAfterToAlternate Then
                        locShiftChangeFlag = locShiftChangeFlag Xor True
                    End If
                Else
                    If Weekday(Date.FromOADate(locDouble), FirstDayOfWeek.Monday) = -locShiftDaysChangeCount Then
                        locShiftChangeFlag = locShiftChangeFlag Xor True
                    End If
                End If
            End If
        Next
    End Sub

    ''' <summary>
    ''' Erstellt einen neue Produktionsdaten-Auswertungszeitraum auf Basis der angegebenen Parameter.
    ''' </summary>
    ''' <param name="StartDate">Das Anfangsdatum des Auswertungszeitraums (inklusive).</param>
    ''' <param name="EndDate">Das Enddatum des Auswertungszeitraums (inklusive)</param>
    ''' <param name="AlternatingShift1">Die erste alterierende Schicht</param>
    ''' <param name="AlternatingShift2">Die zweite alternierende Schicht</param>
    ''' <param name="ShiftChangeAfterDays">Positiv: Die Anzahl der Tage, nachdem die Schichtnummern getauscht werden. Negativ: Die (negative) Wochentagsnummer, an der die Schichtnummern getauscht werden.</param>
    ''' <remarks></remarks>
    Public Sub New(ByVal StartDate As Date, ByVal EndDate As Date, ByVal AlternatingShift1 As Byte, ByVal AlternatingShift2 As Byte, ByVal ShiftChangeAfterDays As Integer)
        MyBase.new()
        StartDate = StartDate.Date
        EndDate = EndDate.Date
        _StartDate = StartDate
        _EndDate = EndDate
        Dim locShiftDaysChangeCount As Integer = 0
        Dim locShiftChangeFlag As Boolean = False
        For locDouble As Double = StartDate.ToOADate To EndDate.ToOADate
            Dim locProductionPeriodItem As New ProductionPeriodItem(Date.FromOADate(locDouble), CByte(IIf(locShiftChangeFlag, AlternatingShift2, AlternatingShift1)))
            locShiftDaysChangeCount += 1
            If locShiftDaysChangeCount > 0 Then
                If locShiftDaysChangeCount = ShiftChangeAfterDays Then
                    locShiftChangeFlag = locShiftChangeFlag Xor True
                End If
            Else
                If Weekday(Date.FromOADate(locDouble), FirstDayOfWeek.Monday) = -locShiftDaysChangeCount Then
                    locShiftChangeFlag = locShiftChangeFlag Xor True
                End If
            End If
        Next
    End Sub

    Public Sub PrepareProductionDates(ByVal IDSubsidiary As Integer, ByVal IDUser As Integer, ByVal Ticket As Date)
        SPAccess.GetInstance.ProductionData_PrepareProductionDates(IDSubsidiary, IDUser, Ticket, Me)
    End Sub

    Public ReadOnly Property StartDate() As Date
        Get
            Return _StartDate
        End Get
    End Property

    Public ReadOnly Property EndDate() As Date
        Get
            Return _EndDate
        End Get
    End Property

    Public ReadOnly Property RangeDescription() As String
        Get
            If StartDate = EndDate Then
                If CultureInfo.CurrentUICulture.TwoLetterISOLanguageName = "de" Then
                    Return "Am " & StartDate.ToLongDateString
                Else
                    Return "On " & StartDate.ToLongDateString
                End If
            Else
                If CultureInfo.CurrentUICulture.TwoLetterISOLanguageName = "de" Then
                    Return "Von " & StartDate.ToLongDateString & " bis " & EndDate.ToLongDateString
                Else
                    Return "From " & StartDate.ToLongDateString & " to " & EndDate.ToLongDateString
                End If
            End If
        End Get
    End Property

    Public ReadOnly Property StartDateMonthDescription() As String
        Get
            Return StartDate.ToString("MMMM yyyy")
        End Get
    End Property

    Public ReadOnly Property ShiftParameters() As ShiftParameters
        Get
            Return _ShiftParameters
        End Get
    End Property

End Class

Public Structure DateRangeParameter
    Private myDateRangePreset As DateRangePresets
    Private myStartDate As Date
    Private myEndDate As Date
    Private myMonthIntoPast As Integer
    Private myLastWorkingDay As LastWorkingdays

    Public Sub New(ByVal DateRangePreset As DateRangePresets, ByVal LastWorkingDay As LastWorkingdays)
        myLastWorkingDay = LastWorkingDay
        Me.DateRangePreset = DateRangePreset
    End Sub

    Public Sub New(ByVal DateRangePreset As DateRangePresets)
        Me.DateRangePreset = DateRangePreset
    End Sub

    Public Sub New(ByVal DateRangePreset As DateRangePresets, ByVal MonthIntoPast As Integer)
        myMonthIntoPast = MonthIntoPast
        Me.DateRangePreset = DateRangePreset
    End Sub

    Public Sub New(ByVal Startdate As Date, ByVal Enddate As Date)
        myDateRangePreset = DateRangePresets.CustomPeriod
        myStartDate = Startdate
        myEndDate = Enddate
    End Sub

    Public Property LastWorkingday() As LastWorkingdays
        Get
            Return myLastWorkingday
        End Get
        Set(ByVal value As LastWorkingdays)
            myLastWorkingday = value
        End Set
    End Property

    Public Property DateRangePreset() As DateRangePresets
        Get
            Return myDateRangePreset
        End Get
        Set(ByVal value As DateRangePresets)
            Select Case value
                Case DateRangePresets.YesterdayOrLastWorkingDay
                    If Date.Now.Date.DayOfWeek = DayOfWeek.Monday Then
                        If LastWorkingday = LastWorkingdays.Sunday Then
                            myStartDate = Date.Now.Date.AddDays(-1)
                        ElseIf LastWorkingday = LastWorkingdays.Saturday Then
                            myStartDate = Date.Now.Date.AddDays(-2)
                        Else
                            myStartDate = Date.Now.Date.AddDays(-3)
                        End If
                    Else
                        myStartDate = Date.Now.Date.AddDays(-1)
                    End If
                    myEndDate = myStartDate
                Case DateRangePresets.FromStartOfCurrentMonthToNow
                    myStartDate = Dates.FirstDayOfMonth(Date.Now.Date)
                    myEndDate = Date.Now.Date
                Case DateRangePresets.FromStartOfCurrentWeekToNow
                    myStartDate = Dates.MondayOfWeek(Date.Now.Date)
                    myEndDate = Date.Now.Date
                Case DateRangePresets.FromStartToEndOfSpecifiedMonth
                    If myMonthIntoPast = -1 Then
                        myMonthIntoPast = 0
                    End If
                    myStartDate = Dates.FirstDayOfMonth(Date.Now.AddMonths(-MonthIntoPast))
                    myEndDate = Dates.LastDayOfMonth(Date.Now.AddMonths(-MonthIntoPast))
                Case DateRangePresets.LastWeek
                    myStartDate = Dates.MondayOfWeek(Date.Now.Date).AddDays(-7)
                    myEndDate = myStartDate.AddDays(7)
                Case DateRangePresets.Today
                    myStartDate = Date.Now.Date
                    myEndDate = myStartDate
                Case DateRangePresets.SinceYearBeganToNow
                    myStartDate = New Date(Date.Now.Year, 1, 1)
                    myEndDate = Date.Now.Date
                Case DateRangePresets.WeekBeforeLastWeek
                    myStartDate = Dates.MondayOfWeek(Date.Now.Date).AddDays(-14)
                    myEndDate = myStartDate.AddDays(7)
            End Select
            myDateRangePreset = value
        End Set
    End Property

    Public Property MonthIntoPast() As Integer
        Get
            Return myMonthIntoPast
        End Get
        Set(ByVal value As Integer)
            myMonthIntoPast = value
        End Set
    End Property

    Public Property StartDate() As Date
        Get
            Return myStartDate
        End Get
        Set(ByVal value As Date)
            myStartDate = value
        End Set
    End Property

    Public Property EndDate() As Date
        Get
            Return myEndDate
        End Get
        Set(ByVal value As Date)
            myEndDate = value
        End Set
    End Property

    Public Overrides Function ToString() As String
        Dim locString As String = ""
        Select Case myDateRangePreset
            Case DateRangePresets.CustomPeriod
                locString = "Freidefinierter Zeitraum."
            Case DateRangePresets.FromStartOfCurrentMonthToNow
                locString = "Vom Anfang des aktuellen Monats bis heute."
            Case DateRangePresets.FromStartOfCurrentWeekToNow
                locString = "Vom Anfang der aktuellen Woche bis heute."
            Case DateRangePresets.FromStartToEndOfSpecifiedMonth
                locString = "Vom Anfang bis zum Ende des Monats " & StartDate.ToString("MMMM yyyy") & "."
            Case DateRangePresets.LastWeek
                locString = "Die letzte Woche."
            Case DateRangePresets.SinceYearBeganToNow
                locString = "Von Anfang des laufenden Jahres bis heute."
            Case DateRangePresets.Today
                locString = "Heute."
            Case DateRangePresets.WeekBeforeLastWeek
                locString = "Die vorletzte Woche."
            Case DateRangePresets.YesterdayOrLastWorkingDay
                locString = "Gestern."
        End Select
        locString &= vbNewLine
        If StartDate = EndDate Then
            locString &= "Aus heutiger Sicht ist das am " & StartDate.ToLongDateString
        Else
            locString &= "Aus heutiger Sicht ist das vom " & StartDate.ToLongDateString & " bis " & EndDate.ToLongDateString
        End If
        Return locString
    End Function
End Structure

Public Enum LastWorkingdays
    Friday
    Saturday
    Sunday
End Enum

Public Enum DateRangePresets
    CustomPeriod = 0
    FromStartOfCurrentMonthToNow
    FromStartToEndOfSpecifiedMonth
    SinceYearBeganToNow
    FromStartOfCurrentWeekToNow
    LastWeek
    YesterdayOrLastWorkingDay
    Today
    WeekBeforeLastWeek
End Enum

File: src/MixNetFx/Facesso.Data/DataClasses/ParameterClasses/ProductionPeriods.cs
using System;
using System.Collections.ObjectModel;
using System.Globalization;
using ActiveDev;

namespace Facesso.Data
{
    public class ProductionPeriod : Collection<ProductionPeriodItem>
    {
        private DateTime _StartDate;
        private DateTime _EndDate;
        private ShiftParameters _ShiftParameters;

        public ProductionPeriod(DateTime certainDate, byte certainShift) : base()
        {
            certainDate = certainDate.Date;
            _StartDate = certainDate;
            _EndDate = certainDate;
            this.Add(new ProductionPeriodItem(certainDate, certainShift));
        }

        /// <summary>
        /// Erstellt einen neuen Produktionsdaten-Auswertungszeitraum über alle Schichten im entsprechenden Zeitraum
        /// </summary>
        public ProductionPeriod(DateTime startdate, DateTime endDate) : base()
        {
            startdate = startdate.Date;
            endDate = endDate.Date;
            _StartDate = startdate;
            _EndDate = endDate;
            for (double locDouble = startdate.ToOADate(); locDouble <= endDate.ToOADate(); locDouble++)
            {
                for (byte locShift = 1; locShift <= 4; locShift++)
                {
                    this.Add(new ProductionPeriodItem(DateTime.FromOADate(locDouble), locShift));
                }
            }
        }

        public ProductionPeriod(DateTime startdate, DateTime endDate, byte shift)
            : this(startdate, endDate, shift, shift, 1) { }

        public ProductionPeriod(DateRangeParameter dateRange, ShiftParameters shifts) : base()
        {
            _StartDate = dateRange.StartDate.Date;
            _EndDate = dateRange.EndDate.Date;
            _ShiftParameters = shifts;

            int locShiftDaysChangeCount = 0;
            bool locShiftChangeFlag = false;
            for (double locDouble = _StartDate.ToOADate(); locDouble <= _EndDate.ToOADate(); locDouble++)
            {
                if (!shifts.AlternateShifts)
                {
                    if (shifts.ConsiderShift1) this.Add(new ProductionPeriodItem(DateTime.FromOADate(locDouble), 1));
                    if (shifts.ConsiderShift2) this.Add(new ProductionPeriodItem(DateTime.FromOADate(locDouble), 2));
                    if (shifts.ConsiderShift3) this.Add(new ProductionPeriodItem(DateTime.FromOADate(locDouble), 3));
                    if (shifts.ConsiderShift4) this.Add(new ProductionPeriodItem(DateTime.FromOADate(locDouble), 4));
                }
                else
                {
                    this.Add(new ProductionPeriodItem(DateTime.FromOADate(locDouble),
                        (byte)(locShiftChangeFlag ? shifts.AlternatingFirstShift : shifts.AlternatingSecondShift)));
                    locShiftDaysChangeCount += 1;
                    if (locShiftDaysChangeCount > 0)
                    {
                        if (locShiftDaysChangeCount == shifts.DaysAfterToAlternate)
                            locShiftChangeFlag = !locShiftChangeFlag;
                    }
                    else
                    {
                        int vbWeekday = VbWeekdayMondayFirst(DateTime.FromOADate(locDouble));
                        if (vbWeekday == -locShiftDaysChangeCount)
                            locShiftChangeFlag = !locShiftChangeFlag;
                    }
                }
            }
        }

        public ProductionPeriod(DateTime startDate, DateTime endDate,
            byte alternatingShift1, byte alternatingShift2, int shiftChangeAfterDays) : base()
        {
            startDate = startDate.Date;
            endDate = endDate.Date;
            _StartDate = startDate;
            _EndDate = endDate;
            int locShiftDaysChangeCount = 0;
            bool locShiftChangeFlag = false;
            for (double locDouble = startDate.ToOADate(); locDouble <= endDate.ToOADate(); locDouble++)
            {
                this.Add(new ProductionPeriodItem(DateTime.FromOADate(locDouble),
                    (byte)(locShiftChangeFlag ? alternatingShift2 : alternatingShift1)));
                locShiftDaysChangeCount += 1;
                if (locShiftDaysChangeCount > 0)
                {
                    if (locShiftDaysChangeCount == shiftChangeAfterDays)
                        locShiftChangeFlag = !locShiftChangeFlag;
                }
                else
                {
                    int vbWeekday = VbWeekdayMondayFirst(DateTime.FromOADate(locDouble));
                    if (vbWeekday == -locShiftDaysChangeCount)
                        locShiftChangeFlag = !locShiftChangeFlag;
                }
            }
        }

        private static int VbWeekdayMondayFirst(DateTime d)
        {
            int dow = (int)d.DayOfWeek;
            return (dow == 0) ? 7 : dow;
        }

        public void PrepareProductionDates(int idSubsidiary, int idUser, DateTime ticket)
        {
            SPAccess.GetInstance().ProductionData_PrepareProductionDates(idSubsidiary, idUser, ticket, this);
        }

        public DateTime StartDate => _StartDate;

        public DateTime EndDate => _EndDate;

        public string RangeDescription
        {
            get
            {
                if (StartDate == EndDate)
                {
                    if (CultureInfo.CurrentUICulture.TwoLetterISOLanguageName == "de")
                        return "Am " + StartDate.ToLongDateString();
                    else
                        return "On " + StartDate.ToLongDateString();
                }
                else
                {
                    if (CultureInfo.CurrentUICulture.TwoLetterISOLanguageName == "de")
                        return "Von " + StartDate.ToLongDateString() + " bis " + EndDate.ToLongDateString();
                    else
                        return "From " + StartDate.ToLongDateString() + " to " + EndDate.ToLongDateString();
                }
            }
        }

        public string StartDateMonthDescription => StartDate.ToString("MMMM yyyy");

        public ShiftParameters ShiftParameters => _ShiftParameters;
    }

    public struct DateRangeParameter
    {
        private DateRangePresets myDateRangePreset;
        private DateTime myStartDate;
        private DateTime myEndDate;
        private int myMonthIntoPast;
        private LastWorkingdays myLastWorkingDay;

        public DateRangeParameter(DateRangePresets dateRangePreset, LastWorkingdays lastWorkingDay)
        {
            myLastWorkingDay = lastWorkingDay;
            myMonthIntoPast = 0;
            myStartDate = default;
            myEndDate = default;
            myDateRangePreset = DateRangePresets.CustomPeriod;
            DateRangePreset = dateRangePreset;
        }

        public DateRangeParameter(DateRangePresets dateRangePreset)
        {
            myLastWorkingDay = LastWorkingdays.Friday;
            myMonthIntoPast = 0;
            myStartDate = default;
            myEndDate = default;
            myDateRangePreset = DateRangePresets.CustomPeriod;
            DateRangePreset = dateRangePreset;
        }

        public DateRangeParameter(DateRangePresets dateRangePreset, int monthIntoPast)
        {
            myMonthIntoPast = monthIntoPast;
            myLastWorkingDay = LastWorkingdays.Friday;
            myStartDate = default;
            myEndDate = default;
            myDateRangePreset = DateRangePresets.CustomPeriod;
            DateRangePreset = dateRangePreset;
        }

        public DateRangeParameter(DateTime startdate, DateTime enddate)
        {
            myDateRangePreset = DateRangePresets.CustomPeriod;
            myStartDate = startdate;
            myEndDate = enddate;
            myMonthIntoPast = 0;
            myLastWorkingDay = LastWorkingdays.Friday;
        }

        public LastWorkingdays LastWorkingday
        {
            get { return myLastWorkingDay; }
            set { myLastWorkingDay = value; }
        }

        public DateRangePresets DateRangePreset
        {
            get { return myDateRangePreset; }
            set
            {
                switch (value)
                {
                    case DateRangePresets.YesterdayOrLastWorkingDay:
                        if (DateTime.Now.Date.DayOfWeek == DayOfWeek.Monday)
                        {
                            if (myLastWorkingDay == LastWorkingdays.Sunday)
                                myStartDate = DateTime.Now.Date.AddDays(-1);
                            else if (myLastWorkingDay == LastWorkingdays.Saturday)
                                myStartDate = DateTime.Now.Date.AddDays(-2);
                            else
                                myStartDate = DateTime.Now.Date.AddDays(-3);
                        }
                        else
                            myStartDate = DateTime.Now.Date.AddDays(-1);
                        myEndDate = myStartDate;
                        break;
                    case DateRangePresets.FromStartOfCurrentMonthToNow:
                        myStartDate = Dates.FirstDayOfMonth(DateTime.Now.Date);
                        myEndDate = DateTime.Now.Date;
                        break;
                    case DateRangePresets.FromStartOfCurrentWeekToNow:
                        myStartDate = Dates.MondayOfWeek(DateTime.Now.Date);
                        myEndDate = DateTime.Now.Date;
                        break;
                    case DateRangePresets.FromStartToEndOfSpecifiedMonth:
                        if (myMonthIntoPast == -1) myMonthIntoPast = 0;
                        myStartDate = Dates.FirstDayOfMonth(DateTime.Now.AddMonths(-MonthIntoPast));
                        myEndDate = Dates.LastDayOfMonth(DateTime.Now.AddMonths(-MonthIntoPast));
                        break;
                    case DateRangePresets.LastWeek:
                        myStartDate = Dates.MondayOfWeek(DateTime.Now.Date).AddDays(-7);
                        myEndDate = myStartDate.AddDays(7);
                        break;
                    case DateRangePresets.Today:
                        myStartDate = DateTime.Now.Date;
                        myEndDate = myStartDate;
                        break;
                    case DateRangePresets.SinceYearBeganToNow:
                        myStartDate = new DateTime(DateTime.Now.Year, 1, 1);
                        myEndDate = DateTime.Now.Date;
                        break;
                    case DateRangePresets.WeekBeforeLastWeek:
                        myStartDate = Dates.MondayOfWeek(DateTime.Now.Date).AddDays(-14);
                        myEndDate = myStartDate.AddDays(7);
                        break;
                }
                myDateRangePreset = value;
            }
        }

        public int MonthIntoPast
        {
            get { return myMonthIntoPast; }
            set { myMonthIntoPast = value; }
        }

        public DateTime StartDate
        {
            get { return myStartDate; }
            set { myStartDate = value; }
        }

        public DateTime EndDate
        {
            get { return myEndDate; }
            set { myEndDate = value; }
        }

        public override string ToString()
        {
            string locString = "";
            switch (myDateRangePreset)
            {
                case DateRangePresets.CustomPeriod:
                    locString = "Freidefinierter Zeitraum."; break;
                case DateRangePresets.FromStartOfCurrentMonthToNow:
                    locString = "Vom Anfang des aktuellen Monats bis heute."; break;
                case DateRangePresets.FromStartOfCurrentWeekToNow:
                    locString = "Vom Anfang der aktuellen Woche bis heute."; break;
                case DateRangePresets.FromStartToEndOfSpecifiedMonth:
                    locString = "Vom Anfang bis zum Ende des Monats " + StartDate.ToString("MMMM yyyy") + "."; break;
                case DateRangePresets.LastWeek:
                    locString = "Die letzte Woche."; break;
                case DateRangePresets.SinceYearBeganToNow:
                    locString = "Von Anfang des laufenden Jahres bis heute."; break;
                case DateRangePresets.Today:
                    locString = "Heute."; break;
                case DateRangePresets.WeekBeforeLastWeek:
                    locString = "Die vorletzte Woche."; break;
                case DateRangePresets.YesterdayOrLastWorkingDay:
                    locString = "Gestern."; break;
            }
            locString += Environment.NewLine;
            if (StartDate == EndDate)
                locString += "Aus heutiger Sicht ist das am " + StartDate.ToLongDateString();
            else
                locString += "Aus heutiger Sicht ist das vom " + StartDate.ToLongDateString() + " bis " + EndDate.ToLongDateString();
            return locString;
        }
    }

    public enum LastWorkingdays
    {
        Friday,
        Saturday,
        Sunday
    }

    public enum DateRangePresets
    {
        CustomPeriod = 0,
        FromStartOfCurrentMonthToNow,
        FromStartToEndOfSpecifiedMonth,
        SinceYearBeganToNow,
        FromStartOfCurrentWeekToNow,
        LastWeek,
        YesterdayOrLastWorkingDay,
        Today,
        WeekBeforeLastWeek
    }
}
